Lib/glob.py

```diff
@@ -310,3 +158,238 @@ def translate(pat, *, recursive=False, include_hidden=False, seps=None):
             results.append(any_sep)
     res = ''.join(results)
     return fr'(?s:{res})\Z'
+
+
+def _compile_pattern(pattern, recursive, include_hidden):
+    regex = translate(pattern, recursive=recursive,
+                      include_hidden=include_hidden, seps=os.path.sep)
+    return re.compile(regex, flags=_pattern_flags).match
+
+
+def _split_pathname(pathname):
+    """Split the given path into a pair (anchor, parts), where *anchor* is the
+    path drive and root (if any), and *parts* is a tuple of path components.
+    """
+    parts = []
+    split = os.path.split
+    dirname, part = split(pathname)
+    while dirname != pathname:
+        parts.append(part)
+        pathname = dirname
+        dirname, part = split(pathname)
+    parts.reverse()
+    return dirname, tuple(parts)
+
+
+def _add_trailing_slash(pathname):
+    """Returns the given path with a trailing slash added, where possible.
+    """
+    tail = os.path.splitdrive(pathname)[1]
+    if not tail or tail[-1] in _path_seps:
+        return pathname
+    return pathname + os.path.sep
+
+
+def _open_dir(path, rel_path, dir_fd):
+    """Prepares the directory for scanning. Returns a 3-tuple with parts:
+
+    1. A path or fd to supply to `os.scandir()`.
+    2. The file descriptor for the directory, or None.
+    3. Whether the caller should close the fd (bool).
+    """
+    if dir_fd is None:
+        return path, None, False
+    elif rel_path == './':
+        return dir_fd, dir_fd, False
+    else:
+        fd = os.open(rel_path, _dir_open_flags, dir_fd=dir_fd)
+        return fd, fd, True
+
+
+@functools.lru_cache(maxsize=1024)
```

Comment:
I pulled 1024 out of thin air, because I don't know how to determine a good value.

Reply:
Using `/usr/bin/time -v` to measure the maximum resident set size, and varying the size of the cache, I arrive at:

- `_literal_selector()`: 0.63 kB per entry
- `_wildcard_selector()`: 0.85 kB per entry
- `_recursive_selector()`: 1.33 kB per entry

So a full `_selector()` cache would consume ~1MB of memory. That seems quite high to me, perhaps I should bring `maxsize` down to 256 or so?

Reply:
I've revised this so that only the `re.Pattern` objects are cached.